pkg/compact/compact.go

```diff
@@ -921,17 +921,22 @@ func (c *BucketCompactor) Compact(ctx context.Context) error {
 	// Loop over bucket and compact until there's no work left.
 	for {
 		var (
-			errGroup, errGroupCtx = errgroup.WithContext(ctx)
-			groupChan             = make(chan *Group)
-			finishedAllGroups     = true
-			mtx                   sync.Mutex
+			wg                     sync.WaitGroup
+			workCtx, workCtxCancel = context.WithCancel(ctx)
+			groupChan              = make(chan *Group)
+			errChan                = make(chan error, c.concurrency)
+			finishedAllGroups      = true
+			mtx                    sync.Mutex
 		)
 
 		// Set up workers who will compact the groups when the groups are ready.
+		// They will compact available groups until they encounter an error, after which they will stop.
 		for i := 0; i < c.concurrency; i++ {
-			errGroup.Go(func() error {
+			go func() {
+				wg.Add(1)
```

Comment:
Could we move this to `main` goroutine as per docs:

> Typically this means the calls to Add should execute before the statement creating the goroutine or other event to be waited for.

https://golang.org/pkg/sync/#WaitGroup.Add

Reply:
Yes! That's a major one. Good spotting :+1: 

Otherwise let's merge.

Reply:
Good catch!  Fixed now